### ISSUE TITLE
Weird type casting issues with Raw and bitwise AND (& [ampersand])

### ISSUE THREAD

**Issue by zjr** [tgriesser/knex#3346]
# Environment

Knex version: 0.19.0
Database + version: PostgreSQL v11.3
OS: MacOS 10.14.5

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do

Would like to use bitwise AND operator (`&`) in a query to PostgreSQL to but something seems to casting my bits to strings.

2. Error message

```
error: operator is not unique: unknown & unknown
    at Connection.parseE (/Users/zjr/Projects/top-notch/node_modules/pg/lib/connection.js:602:11)
    at Connection.parseMessage (/Users/zjr/Projects/top-notch/node_modules/pg/lib/connection.js:399:19)
    at Socket.<anonymous> (/Users/zjr/Projects/top-notch/node_modules/pg/lib/connection.js:121:22)
    at Socket.emit (events.js:200:13)
    at addChunk (_stream_readable.js:294:12)
    at readableAddChunk (_stream_readable.js:275:11)
    at Socket.Readable.push (_stream_readable.js:210:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:166:17) {
  name: 'error',
  length: 190,
  severity: 'ERROR',
  code: '42725',
  detail: undefined,
  hint: 'Could not choose a best candidate operator. You might need to add explicit type casts.',
  position: '35',
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'parse_oper.c',
  line: '720',
  routine: 'op_error'
}
```

3. Reduced test code, for example in https://npm.runkit.com/knex or if it needs real
   database connection to MySQL or PostgreSQL, then single file example which initializes
   needed data and demonstrates the problem.

```js
const Knex = require('knex');
const knex = Knex({ client: 'pg' });

knex('foo').where(knex.raw('? & ? <> 0', [2, 1])).then(console.log).catch(console.error);
```

Here's an example on run kit if it's enlightening at all: https://runkit.com/zjr/5d27b06e58237b00181eb044.

As a work-around in the meantime I'm doing this: `knex.raw('?::integer & ?::integer <> 0', [2, 1])` and that seems to work. It also works if you just do `knex.raw('2 & 1 <> 0')`, if that makes a difference.

I realize this might be an issue with `pg` and not `knex` but I thought it made sense to start here.

**Comment by elhigu** [comment]
Looks like parameters of prepared statements doesn't have the same type information available, that plain sql.

This is what knex sends to node-pg driver: (output of `.toSQL().toNative()`)

```
{bindings: [2, 1], sql: "select \"foo\".* from \"foo\" as \"f\", \"bar\" as \"b\" where $1 & $2 <> 0"}
```

Which looks as expected. There is nothing that knex can do about this and I don't know exactly why postgresql behaves like that when accessed using prepared statements and parameter bindings with node-pg.

**Comment by elhigu** [comment]
My best guess is that when pg driver sends statement to be prepared it doesn't know types of the parameter bindings and is not able to prepare that operator 🤔 When same stuff is sent as SQL string DB knows both parameter types already when interpreting query.

**Comment by zjr** [comment]
Great, thanks for the quick & thoughtful response!

Please pardon my naiveté but is this something I should open an issue for at node-postgres?

**Comment by elhigu** [comment]
Could be that @brianc can confirm if my theory is correct.

If it is correct, then there is nothing that can make it to work automatically without casting, unless we provide parameter types to prepared statements. Though I don't now if node-pg supports that and in that case also you would need to tell parameter types manually.

tl;dr there is really no way to make it work better unless we stop using prepared statements to execute queries (that is insecure, so I we will not do that) .
